### ISSUE TITLE
Corrupted output writing large array to an AnonymousPipeServerStream on linux using netcoreapp3.1 

### ISSUE THREAD

**Issue by NegativeMjark** [dotnet/runtime#70798]
### Description

Writing a large array to an AnonymousPipeServerStream on linux sometimes writes the wrong output. I think the problem is in `PipeStream.WriteCore` 
https://github.com/dotnet/runtime/blob/v5.0.17/src/libraries/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs#L169-L177
where if the syscall doesn't write the entire array in one go the PipeStream tries to write the remainder of the array, however it doesn't update the pointer so starts writing from the start of the array.
```
            // For anonymous pipes, write the file descriptor.
            fixed (byte* bufPtr = &MemoryMarshal.GetReference(buffer))
            {
                while (buffer.Length > 0)
                {
                    int bytesWritten = CheckPipeCall(Interop.Sys.Write(_handle, bufPtr, buffer.Length));
                    buffer = buffer.Slice(bytesWritten);
                }
            }
```

### Reproduction Steps

Write a large byte[] array to an AnonymousPipeServerStream on linux on netcoreapp3.1

### Expected behavior

That it writes the given byte array to the stream.

### Actual behavior

It sometimes writes a different sequence of bytes to the stream.

### Regression?

This seems to be introduced by https://github.com/dotnet/corefx/commit/5bac18d126be466cf9ea01d0132f31fbe81edae4#diff-b34cfcf05d3ed7dddff30ea4fc1ddcf1986fd2293801aa9ed96c362b769acddaR153 so probably affects most of the versions before net6

### Known Workarounds

It looks like the code was changed for net6 by https://github.com/dotnet/runtime/commit/2606d828b854aa13572568ebd4798ee6860f7107
The net6 version of the code doesn't seem to have the same issue.

On versions prior to net6 writing the array in smaller chunks seems to be less likely to trigger this bug.

**Comment by stephentoub** [comment]
Yes, this was fixed in .NET 6.

**Comment by adamsitnik** [comment]
@stephentoub do you think that it would meet the 3.1 backport bar?

**Comment by danmoseley** [comment]
> @stephentoub do you think that it would meet the 3.1 backport bar?

.NET 3.1 is in its[ final 6 months](https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-and-net-core) of support. We call that the 'maintenance period' and we [only take security fixes in that period.](https://github.com/dotnet/core/blob/main/release-policies.md#maintenance-support) 

@NegativeMjark is it possible to update to .NET 6 to avoid this bug?

**Comment by NegativeMjark** [comment]
For my use case we'll need to update to .NET 6 at some point anyway. In the meantime writing the array in small page-sized chunks seems to work as a mitigation. So I don't think we urgently need a fix for 3.1.

If you did want to fix it for 3.1 I've put a 2-line PR together at https://github.com/dotnet/corefx/pull/43149

**Comment by jeffhandley** [comment]
Thanks, @NegativeMjark. Since you don't urgently need a fix for 3.1, I'm going to close this issue and your PR. Thanks for your efforts on this.
